### ISSUE TITLE
NFS mounts fail with 1.7.0 and 1.7.1 on Mavericks

### ISSUE THREAD

**Issue by posita** [Parallels/boot2docker-vagrant-box#14]
I'm using the following in my 'Vagrantfile':

```rb
require('fileutils')

Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
    config.vm.define(:boot2docker) do |boot2docker|
        # ...
        boot2docker.vm.synced_folder('.', '/vagrant', disabled: true)
        mnt_path = FileUtils.pwd()
        boot2docker.vm.synced_folder(mnt_path, mnt_path, nfs: true)
        # ...
    end
end
```

1.6.0 works fine for `nfs` mounts:

```sh
% uname -a
Darwin [...] 13.4.0 Darwin Kernel Version 13.4.0: Wed Mar 18 16:20:14 PDT 2015; root:xnu-2422.115.14~1/RELEASE_X86_64 x86_64
% vagrant up
Bringing machine 'boot2docker' up with 'parallels' provider...
==> boot2docker: Importing base box 'parallels/boot2docker'...
==> boot2docker: Checking if box 'parallels/boot2docker' is up to date...
==> boot2docker: A newer version of the box 'parallels/boot2docker' is available! You currently
==> boot2docker: have version '1.6.0'. The latest is version '1.7.1'. Run
==> boot2docker: `vagrant box update` to update.
==> boot2docker: Setting the name of the VM: boot2docker
==> boot2docker: Preparing network interfaces based on configuration...
    boot2docker: Adapter 0: shared
    boot2docker: Adapter 1: hostonly
==> boot2docker: Clearing any previously set network interfaces...
==> boot2docker: Setting power consumption mode: "Longer battery life"
==> boot2docker: Running 'pre-boot' VM customizations...
==> boot2docker: Booting VM...
==> boot2docker: Waiting for machine to boot. This may take a few minutes...
    boot2docker: SSH address: 10.211.55.194:22
    boot2docker: SSH username: docker
    boot2docker: SSH auth method: password
    boot2docker:
    boot2docker: Inserting generated public key within guest...
    boot2docker: Removing insecure key from the guest if its present...
    boot2docker: Key inserted! Disconnecting and reconnecting using new SSH key...
==> boot2docker: Machine booted and ready!
==> boot2docker: Configuring and enabling network interfaces...
==> boot2docker: Exporting NFS shared folders...
==> boot2docker: Preparing to edit /etc/exports. Administrator privileges will be required...
Password:
==> boot2docker: Mounting NFS shared folders...
==> boot2docker: Running triggers after up...
...
```

This is what I get for 1.7.0:

```sh
Bringing machine 'boot2docker' up with 'parallels' provider...
==> boot2docker: Importing base box 'parallels/boot2docker'...
==> boot2docker: Checking if box 'parallels/boot2docker' is up to date...
==> boot2docker: A newer version of the box 'parallels/boot2docker' is available! You currently
==> boot2docker: have version '1.7.0'. The latest is version '1.7.1'. Run
==> boot2docker: `vagrant box update` to update.
==> boot2docker: Setting the name of the VM: boot2docker
==> boot2docker: Preparing network interfaces based on configuration...
    boot2docker: Adapter 0: shared
    boot2docker: Adapter 1: hostonly
==> boot2docker: Clearing any previously set network interfaces...
==> boot2docker: Setting power consumption mode: "Longer battery life"
==> boot2docker: Running 'pre-boot' VM customizations...
==> boot2docker: Booting VM...
==> boot2docker: Waiting for machine to boot. This may take a few minutes...
    boot2docker: SSH address: 10.211.55.195:22
    boot2docker: SSH username: docker
    boot2docker: SSH auth method: password
    boot2docker:
    boot2docker: Inserting generated public key within guest...
    boot2docker: Removing insecure key from the guest if its present...
    boot2docker: Key inserted! Disconnecting and reconnecting using new SSH key...
==> boot2docker: Machine booted and ready!
==> boot2docker: Checking for Parallels Tools installed on the VM...
==> boot2docker: Parallels Tools were not detected on this VM! They are required
==> boot2docker: for forwarded ports, shared folders, host only networking and more.
==> boot2docker: If SSH fails or shared folders are not working on this machine,
==> boot2docker: please install Parallels Tools within the virtual machine and
==> boot2docker: reload your VM.
==> boot2docker: Configuring and enabling network interfaces...
==> boot2docker: Exporting NFS shared folders...
==> boot2docker: Preparing to edit /etc/exports. Administrator privileges will be required...
==> boot2docker: Mounting NFS shared folders...
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

mount -o 'vers=3,udp' 10.211.55.2:'/Users/[path/to/vagrant/directory]' /Users/Users/[path/to/vagrant/directory]

Stdout from the command:



Stderr from the command:

mount.nfs: an incorrect mount option was specified
mount: mounting 10.211.55.2:/Users/Users/[path/to/vagrant/directory] on /Users/Users/[path/to/vagrant/directory] failed: Connection refused
```

I get the same for 1.7.1 (without the box version warning):

```sh
Bringing machine 'boot2docker' up with 'parallels' provider...
==> boot2docker: Importing base box 'parallels/boot2docker'...
==> boot2docker: Checking if box 'parallels/boot2docker' is up to date...
==> boot2docker: Setting the name of the VM: boot2docker
==> boot2docker: Preparing network interfaces based on configuration...
    boot2docker: Adapter 0: shared
    boot2docker: Adapter 1: hostonly
==> boot2docker: Clearing any previously set network interfaces...
==> boot2docker: Setting power consumption mode: "Longer battery life"
==> boot2docker: Running 'pre-boot' VM customizations...
==> boot2docker: Booting VM...
==> boot2docker: Waiting for machine to boot. This may take a few minutes...
    boot2docker: SSH address: 10.211.55.196:22
    boot2docker: SSH username: docker
    boot2docker: SSH auth method: password
    boot2docker: Warning: Connection refused. Retrying...
    boot2docker:
    boot2docker: Inserting generated public key within guest...
    boot2docker: Removing insecure key from the guest if its present...
    boot2docker: Key inserted! Disconnecting and reconnecting using new SSH key...
==> boot2docker: Machine booted and ready!
==> boot2docker: Checking for Parallels Tools installed on the VM...
==> boot2docker: Parallels Tools were not detected on this VM! They are required
==> boot2docker: for forwarded ports, shared folders, host only networking and more.
==> boot2docker: If SSH fails or shared folders are not working on this machine,
==> boot2docker: please install Parallels Tools within the virtual machine and
==> boot2docker: reload your VM.
==> boot2docker: Configuring and enabling network interfaces...
==> boot2docker: Exporting NFS shared folders...
==> boot2docker: Preparing to edit /etc/exports. Administrator privileges will be required...
==> boot2docker: Mounting NFS shared folders...
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

mount -o 'vers=3,udp' 10.211.55.2:'/Users/[path/to/vagrant/directory]' /Users/[path/to/vagrant/directory]

Stdout from the command:



Stderr from the command:

mount.nfs: an incorrect mount option was specified
mount: mounting 10.211.55.2:/Users/[path/to/vagrant/directory] on /Users/[path/to/vagrant/directory] failed: Connection refused
```

Changing the first argument to `synched_folder` to `'.'` and removing the line `boot2docker.vm.synced_folder('.', '/vagrant', disabled: true)` doesn't seem help any. (I get the same results.)

Logging in via `vagrant ssh` and trying to `mount` explicitly gives the same error:

```sh
% cat /etc/exports
# VAGRANT-BEGIN: 501 5c733968-21bb-4716-b43e-56252f575c97
"/Users/[path/to/vagrant/directory]" 10.211.55.199 -alldirs -mapall=501:501
# VAGRANT-END: 501 5c733968-21bb-4716-b43e-56252f575c97
% vagrant ssh
...
docker@boot2docker:~$ sudo mkdir /mnt/foo
docker@boot2docker:~$ sudo mount 10.211.55.2:/Users/[path/to/vagrant/directory] /mnt/foo
mount.nfs: an incorrect mount option was specified
mount: mounting 10.211.55.2:/Users/[path/to/vagrant/directory] on /mnt/foo failed: Connection refused
docker@boot2docker:~$ sudo mount -t nfs -o vers=3,udp 10.211.55.2:/Users/[path/to/vagrant/directory] /mnt/foo
mount.nfs: an incorrect mount option was specified
mount: mounting 10.211.55.2:/Users/[path/to/vagrant/directory] on /mnt/foo failed: Connection refused
```

**Comment by posita** [comment]
FYI, I'm noticing the same things with [dduportal/boot2docker-vagrant-box](/dduportal/boot2docker-vagrant-box) running on `virtualbox`, meaning NFS works in 1.6.0 and not in 1.7.1 (very similar symptoms as described above).

**Comment by legal90** [comment]
@posita I've figured out that it could be worked around with specifying "nolock" mount option:
```ruby
config.vm.synced_folder(mnt_path, mnt_path, nfs: true, mount_options: ["nolock"])
```

It was discussed in https://github.com/dduportal/boot2docker-vagrant-box/issues/10#issuecomment-77740430

@dduportal Have you figured out why it doesn't work without "nolock" option?

**Comment by dduportal** [comment]
Hi !

Nope, i have to admit that I completely trusted contributors in that.
I started to read a bit of doc on the implication to the nolock option, but it appears to be used in 2 cases :
* Production performances issues,
* Older NFS server compatibility.

Considering that :
* We are in development use case (or you just do not use vagrant at all if production is targeted)
* Files in the shared folder should not been written with concurrency considering Docker paradigms (use data volumes, with single shot copying, or build you images by sending data thru the Docker daemon API inside images)

So i figured out to accept an Oob working solution, and if problems would have benne appeared with this option, i had to 
* Implement a test around that problem
* Understand the implications helped by community
* Correct that, even if choice has to be made.

Does the use of this option causes problem for you for now ? In the future (i don't have your insights in mind, maybe i miss use cases :) ) ?

**Comment by posita** [comment]
*TL;DR: ideally `nolock` should be set by the administrator on a case-by-case basis, it shouldn't be a blanket requirement if at all possible.*

@legal90, thanks for the work-around! :grinning: Unfortunately, as @dduportal points out, `nolock` should likely be an *option*, not a requirement (I don't think supporting ancient NFS implementations applies here). Even where you need more performance, you still have to make sure you won't need FS-level locking, which I'm not sure you can guarantee with all host mounts.

That being said, many applications seem to have their own locking mechanisms that accommodate NFS's `nolock`. MySQL, for example, has the `--external-locking` and `--skip-external-locking` options (external locking has apparently been [turned off by default since 4.0](https://dev.mysql.com/doc/refman/5.7/en/server-options.html#option_mysqld_external-locking)). Running two [`bitcoind`](/bitcoin/bitcoin) instances syncing blockchain data on a host-mounted file system seems to do the right thing (the later one refuses to start since it can't get a lock on the data directory, although I didn't start them anywhere near the same time, so I'm not confident everything works in all cases).

**Comment by dduportal** [comment]
@posita that's right in a platform mind, we are here in a development environment, given that you should not run applications based on the data which is on this filesystem.
If you want to simulate a production ready environment, use a container for that : you'll be full controlling the NFS server configuration of this container .

I really think that shared folder should be used in one shot mode, so the ```nolock``` should not be a problem.

I'm a really big fan of using rsync for shared folder : it will make your workflow a "push" one and will avoid the sharing problems (exemple symlink with a Windows host, user rights with Mac OS which have strange UIDs, etc.).

**Comment by posita** [comment]
Looks like I spoke too soon regarding a work-around. :disappointed: When using a 1.7.1 release, despite using NFS with `nolock` on all mounts, after some number of repetitions of `vagrant halt ; vagrant up`, I am running into Level/levelup#222. :confused:

So there's something wonky about mounts after 1.6.0 that I don't think is addressed via `nolock` (in other words, `nolock` may be a red herring). I'm trying to create a small testcase to reproduce, but it's difficult, since the symptoms don't appear right away.

@dduportal, with respect to your comments regarding dev environments, best practices, etc. I'm less confident that your proposed limitations accommodate legitimate use cases. Here's one example:

1. Developer runs OS X on a laptop.
1. Production runs `bitcoind` in Docker containers.
1. As of this writing, `bitcoind` requires downloading 50+ GB of blockchain data (which can take days) before being usable.
1. Developer uses Vagrant and boot2docker for development and testing `bitcoind` container.
1. Developer wants blockchain data to persist beyond Vagrant life cycle (i.e., `vagrant destroy` is a necessary part of the development process).

Constraints:

1. No hardware (either new machines or accessories) may be added to the Developer's setup.

The (perhaps naive) solution that comes to mind is store the blockchain data on the native file system, mount it via Vagrant in boot2docker and then use Docker data containers to mount the boot2docker mount points inside the appropriate container. (By the way, I'm not the only person to think of this. This seems to be the [recommended method](https://docs.docker.com/installation/mac/#mount-a-volume-on-the-container) when using boot2docker on a Mac, whether or not Vagrant is involved. That being said, Vagrant affords some administrative conveniences that I'm personally not yet ready to abandon.)

In the above example, my naive solution works without issue in 1.6.0. It fails after a number of Vagrant cycles starting in 1.7.0 (again, after some playing, I think the `nolock` thing probably has nothing to do with the failure).

**Comment by posita** [comment]
Okay, setting aside the `nolock` issue for a moment, I created a [convoluted test-ish thingy](/posita/vagrant-b2d-nfs-test) to explore various combinations. What I found was that [Parallels/boot2docker-vagrant-box](/Parallels/boot2docker-vagrant-box) differs substantially from [dduportal/boot2docker-vagrant-box](/dduportal/boot2docker-vagrant-box) in what it can accommodate.

Three main observations:

1. Parallels/boot2docker-vagrant-box 1.7.1 with `nolock` has permission problems when writing to NFS mounts (see [this](/posita/vagrant-b2d-nfs-test#16x-with-17x-installed-users-now-fails-to-mount-after-vagrant-halt-confused)).
2. Parallels/boot2docker-vagrant-box 1.7.1 is installed alongside 1.6.0, and you use 1.6.0 explicitly (e.g., via `config.vm.box_version`), mounts break until you uninstall 1.7.1 (see [this](https://github.com/posita/vagrant-b2d-nfs-test#16x-with-17x-installed-users-now-fails-to-mount-after-vagrant-halt-confused)).
3. dduportal/boot2docker-vagrant-box with `nolock` pretty much just works.<sup>1</sup>

--

<sup>1</sup> One minor nitpick of dduportal/boot2docker: it only supports TLS connections to Docker via `127.0.0.1` and Vagrant's dynamically-assigned IP address. In other words, you lose all convenience of setting an explicit IP address in your `Vagrantfile` (e.g., `config.vm.network(:private_network, ip: '10.1.2.3')`).

**Comment by posita** [comment]
Okay, setting aside the `nolock` debate for a moment, I created a [convoluted test-ish thingy](/posita/vagrant-b2d-nfs-test) to explore various combinations. What I found was that [Parallels/boot2docker-vagrant-box](/Parallels/boot2docker-vagrant-box) differs substantially from [dduportal/boot2docker-vagrant-box](/dduportal/boot2docker-vagrant-box) in what it can accommodate.

Three main observations:

1. Parallels/boot2docker-vagrant-box 1.7.1 with `nolock` has permission problems when writing to NFS mounts (see [this](/posita/vagrant-b2d-nfs-test#17x-with-nolock-sort-of-works-writing-from-docker-container-doesnt-work-right)).
2. Parallels/boot2docker-vagrant-box 1.7.1 is installed alongside 1.6.0, and you use 1.6.0 explicitly (e.g., via `config.vm.box_version`), mounts break until you uninstall 1.7.1 (see [this](https://github.com/posita/vagrant-b2d-nfs-test#16x-with-17x-installed-users-now-fails-to-mount-after-vagrant-halt-confused)).
3. dduportal/boot2docker-vagrant-box with `nolock` pretty much just works.<sup>1</sup>

--

<sup>1</sup> One minor nitpick of dduportal/boot2docker: it only supports TLS connections to Docker via `127.0.0.1` and Vagrant's dynamically-assigned IP address. In other words, you lose all convenience of setting an explicit IP address in your `Vagrantfile` (e.g., `config.vm.network(:private_network, ip: '10.1.2.3')`).

**Comment by legal90** [comment]
> 2. Parallels/boot2docker-vagrant-box 1.7.1 is installed alongside 1.6.0, and you use 1.6.0 explicitly (e.g., via config.vm.box_version), mounts break when restarting the VM until you uninstall 1.7.1 (see this).

TL;DR: After switching the box version you should recreate a VM from the scratch (`vagrant destroy && vagrant up`)

On the second step of your test you have switched the box version from "1.6.0" to "1.7.1". Each box version has its own Vagrantfile, bundled with the box. Vagrant always parsed it according to the load order(https://docs.vagrantup.com/v2/vagrantfile/index.html).

So, Vagrantiles in `parallels/boot2docker` v1.6.0 and v1.7.1 are different:
v1.6.0: https://github.com/Parallels/boot2docker-vagrant-box/blob/parallels/v1.6.0/vagrantfile.tpl
v1.7.1: https://github.com/Parallels/boot2docker-vagrant-box/blob/parallels/v1.7.1/vagrantfile.tpl

In the v1.7.1 we are mounting `/Users` via Parallels Shared Folders (`prl_fs`), not via NFS. In v1.6.0 `prl_fs` is not supported.
Yet another important thing: starting since v1.7.0 the boot2docker.iso is dumped to the virtual machine HDD. It simplifies the box (we don't have to lookup and attach an iso), but it means that if we switch the box version for the existing VM, then we have to destroy&up this VM to take an effect (like any other guests for Vagrant).

**Comment by posita** [comment]
> TL;DR: After switching the box version you should recreate a VM from the scratch (`vagrant destroy && vagrant up`)

This was my mistake. I was operating under the misunderstanding that a Vagrant instance was pegged to the box+version, not just the box. I have fixed this in a [branch from my original tests](https://github.com/posita/vagrant-b2d-nfs-test/tree/posita-fix-version), and so long as I explicitly set `config.vm.box_version` in the `Vagrantfile`, it doesn't matter which version(s) are installed.

**Comment by posita** [comment]
>> Parallels/boot2docker-vagrant-box 1.7.1 with `nolock` has permission problems when writing to NFS mounts (see [this](/posita/vagrant-b2d-nfs-test#17x-with-nolock-sort-of-works-writing-from-docker-container-doesnt-work-right)).

> In the v1.7.1 we are mounting `/Users` via Parallels Shared Folders (`prl_fs`), not via NFS. In v1.6.0 `prl_fs` is not supported.

If I explicitly disable the `prl_fs` `/Users` mount, [my test works](/posita/vagrant-b2d-nfs-test/tree/posita-disable-users#17x-with-nolock-works-but-only-if-default-users-mount-is-explicitly-disabled). It looks like this is because shared folders are mounted *after* NFS shared folders, so even if `${HOME}` is mounted via NFS, it will be hidden when `/Users` is later mounted over it via `prl_fs`.

While this is different from @dduportal's VirtualBox behavior, I'm no longer confident this should be classified as a bug. The default behaviors in [Parallels/boot2docker-vagrant-box](/Parallels/boot2docker-vagrant-box) caused problems (or rather, behavioral discrepancies) for my particular configuration, but those can be addressed by additional settings in one's `Vagrantfile`.

Apologies for any inconvenience.

**Comment by dduportal** [comment]
Hello !

Thanks for sharing your work and feedback, I understand better use cases and the problem you ran trough. It was tricky !


To let you know, i am thinking on moving to docker-machine since the 0.4.0(.*) seems to bring a lot of new useful features (daemon config, KVP for env like http_proxy), Windows support. The only "problem" left is compose on Windows.

In my company, we have forked the dduportal/boot2docker to address hundred of developers, and we are seriously thinking doing the move on docker-machine to leverage the maintaining effort. I think that it would be the same for the image.

What is your thought about that ?

I would be interested to draw some kind of matrix to help people choose between vagrant stack / docker machine stack to see pros ans cons, would you be interested on that ?

**Comment by legal90** [comment]
@dduportal Docker Machine is really cute and it is a much more convenient way to manage Docker hosts rather than Vagrant. But it will not have a Parallels support in v0.4.0 due to our PR has been declined (https://github.com/docker/machine/pull/939#issuecomment-126818335).

Seems like we have to support this vagrant box until Docker Machine v0.5.0 with the new plugin model is released. We will also publish the custom `docker-machine_darwin-amd64` binary with Parallels driver support, but not all customers will be satisfied by this solution (https://github.com/docker/machine/pull/939#issuecomment-129480266).

**Comment by dduportal** [comment]
Just read that, yeah it's not easy food, but i think you're in the right direction !

Thank for feedback !

**Comment by posita** [comment]
I'd love to be able to move to machine, but likely won't until Docker provides some avenue to use Parallels. I like the flexibility of Vagrant from a dev/admin perspective (complexities can be encapsulated in a `Vagrantfile` and devs can run `vagrant up` to get a working environment). There will probably be ways to do the same thing with Docker Machine (wrapper scripts?), but I won't likely look into them until that team gets its act together. (At least) twice in past few weeks, different Docker teams have [ripped the carpet out from under their users without any warning](https://github.com/docker/hub-beta-feedback/issues/211). I hope those are just growing pains, and not indicative of a more engrained culture.... :disappointed:

**Comment by dduportal** [comment]
Without jumping in the debates which "local optimum or global optimum", I think your point is right given that vagrant as 5 years of life at least, since docker-machine is in 0.4.0-release candidate :)

The target is 2 months at least, given they make a monthly version update, and you need 0.5.0 with the new plugin architecure that will allow your code to be merged.

Keep in mind that Vagrant is dev. oriented, so you create a dev/ops wall (but maybe it will fit to your organisation so it would not be a problem !), but Docker try to address a "no-wall" world : provisionning your infra is basically docker run or docker-compose up in their ideas : high level reproductiblity so on.

So you get it right : the workflow may be slightly different. If it does not feet, do not hesitate to replace Docker (I used packer + vagrant since a long tiime and it was working), given Docker may add much more complexity :)

Whatever, as @legal90 stated, for the next months, vagrant box should be maintained while docker-machien reach its goals.

Thank you for the feedbacks !
